yjit_core.c

```diff
@@ -172,20 +179,63 @@ void ctx_set_opnd_type(ctx_t* ctx, insn_opnd_t opnd, val_type_t type)
     switch (mapping.kind)
     {
         case TEMP_SELF:
-        ctx->self_type = type;
+        UPGRADE_TYPE(ctx->self_type, type);
         break;
 
         case TEMP_STACK:
-        ctx->temp_types[ctx->stack_size - 1 - opnd.idx] = type;
+        {
+            int stack_index = ctx->stack_size - 1 - opnd.idx;
+            UPGRADE_TYPE(ctx->temp_types[stack_index], type);
+        }
         break;
 
         case TEMP_LOCAL:
         RUBY_ASSERT(mapping.idx < MAX_LOCAL_TYPES);
-        ctx->local_types[mapping.idx] = type;
+        UPGRADE_TYPE(ctx->local_types[mapping.idx], type);
         break;
     }
 }
 
+void
+ctx_swap_opnd_mapping(ctx_t* ctx, insn_opnd_t opnd0, insn_opnd_t opnd1)
+{
+    // temp stack only for now
+    RUBY_ASSERT(!opnd0.is_self);
+    RUBY_ASSERT(!opnd1.is_self);
+
+    RUBY_ASSERT(opnd0.idx < ctx->stack_size);
+    RUBY_ASSERT(opnd1.idx < ctx->stack_size);
+
+    int idx0 = ctx->stack_size - 1 - opnd0.idx;
+    int idx1 = ctx->stack_size - 1 - opnd1.idx;
+
+    val_type_t temp_type = ctx->temp_types[idx0];
+    temp_mapping_t temp_mapping = ctx->temp_mapping[idx0];
+
+    ctx->temp_types[idx0] = ctx->temp_types[idx1];
+    ctx->temp_mapping[idx0] = ctx->temp_mapping[idx1];
+
+    ctx->temp_types[idx1] = temp_type;
+    ctx->temp_mapping[idx1] = temp_mapping;
+}
+
+void
+ctx_copy_opnd_mapping(ctx_t* ctx, insn_opnd_t dest, insn_opnd_t src)
```

Comment:
I would like to see comments explaining what these methods do and why they exist, to help current and future team members understand.